File: sdk/typescript/src/resources.ts
import type { HttpClient } from "./http.ts";
import type { Resolver } from "./resolve.ts";
import type {
  Agent,
  AgentInput,
  AgentPatch,
  Connection,
  ConnectionProvider,
  Environment,
  EnvironmentInput,
  EnvironmentPatch,
  Secret,
  Vault,
  VaultInput,
  VaultPatch,
} from "./types.ts";

/**
 * Payloads here use the API's own key names (`environment_id`, `mcp_servers`,
 * `allowed_vault_ids`) rather than camelCase. That is a choice, not an
 * oversight: an agent definition is the same object in the REST API, in a
 * `fountain.yml` manifest and here, so one definition reads identically in all
 * three and the API reference doubles as the SDK reference. Options that
 * control the SDK's own behaviour — `timeoutMs`, `signal` — are camelCase,
 * because those are not data.
 */

/**
 * List, read, define, change and delete one kind of thing.
 *
 * `Patch` is its own parameter rather than `Partial<Input>`: the API's update
 * schemas are not simply every create field made optional, and the generated
 * ones say so exactly.
 */
class Collection<T extends { id: string }, Input, Patch> {
  protected readonly http: HttpClient;
  protected readonly resolver: Resolver;
  protected readonly path: string;
  protected readonly what: string;

  constructor(http: HttpClient, resolver: Resolver, path: string, what: string) {
    this.http = http;
    this.resolver = resolver;
    this.path = path;
    this.what = what;
  }

  /** Everything on the account. */
  async list(search?: string): Promise<T[]> {
    return this.http.list<T>(this.path, { query: { search } });
  }

  /** One of them, by name or id. */
  async get(nameOrId: string): Promise<T> {
    const { id } = await this.resolver.resolve(this.path, this.what, nameOrId);
    return this.http.data<T>("GET", `${this.path}/${id}`);
  }

  /** Define a new one. */
  async create(input: Input): Promise<T> {
    const created = await this.http.data<T>("POST", this.path, { body: input });
    // A new name has to be findable by the next `run({ agent: "…" })`.
    this.resolver.forget(this.path);
    return created;
  }

  /** Change one. Only the fields you pass are touched. */
  async update(nameOrId: string, patch: Patch): Promise<T> {
    const { id } = await this.resolver.resolve(this.path, this.what, nameOrId);
    const updated = await this.http.data<T>("PATCH", `${this.path}/${id}`, { body: patch });
    // The patch may have renamed it.
    this.resolver.forget(this.path);
    return updated;
  }

  /** Delete one. */
  async delete(nameOrId: string): Promise<void> {
    const { id } = await this.resolver.resolve(this.path, this.what, nameOrId);
    await this.http.request("DELETE", `${this.path}/${id}`);
    this.resolver.forget(this.path);
  }
}

/**
 * Secrets on an environment or a vault.
 *
 * Values are write-only: `list` returns keys, never what they are worth. That
 * is the whole point — the SDK can put a credential into a sandbox and can
 * never read it back out.
 */
class Secrets {
  private readonly http: HttpClient;
  private readonly resolver: Resolver;
  private readonly parentPath: string;
  private readonly what: string;

  constructor(http: HttpClient, resolver: Resolver, parentPath: string, what: string) {
    this.http = http;
    this.resolver = resolver;
    this.parentPath = parentPath;
    this.what = what;
  }

  /** The keys stored here. Never the values. */
  async list(parent: string): Promise<Secret[]> {
    return this.http.list<Secret>(`${await this.parentId(parent)}/secrets`);
  }

  /** Store a secret, or replace one with the same key. */
  async set(parent: string, key: string, value: string): Promise<Secret> {
    return this.http.data<Secret>("POST", `${await this.parentId(parent)}/secrets`, {
      body: { key, value },
    });
  }

  /** Store several at once. */
  async setAll(parent: string, secrets: Record<string, string>): Promise<Secret[]> {
    const base = await this.parentId(parent);
    const out: Secret[] = [];
    // Serially: these are audited writes, and a partial failure should say
    // which key it stopped on rather than leave a scattered half-write.
    for (const [key, value] of Object.entries(secrets)) {
      out.push(
        await this.http.data<Secret>("POST", `${base}/secrets`, { body: { key, value } }),
      );
    }
    return out;
  }

  /** Remove one, by key. */
  async delete(parent: string, key: string): Promise<void> {
    await this.http.request("DELETE", `${await this.parentId(parent)}/secrets/${encodeURIComponent(key)}`);
  }

  private async parentId(nameOrId: string): Promise<string> {
    const { id } = await this.resolver.resolve(this.parentPath, this.what, nameOrId);
    return `${this.parentPath}/${id}`;
  }
}

/** The agents on this account. */
export class Agents extends Collection<Agent, AgentInput, AgentPatch> {
  constructor(http: HttpClient, resolver: Resolver) {
    super(http, resolver, "/api/agents", "agent");
  }
}

/** The environments on this account, and their secrets. */
export class Environments extends Collection<Environment, EnvironmentInput, EnvironmentPatch> {
  readonly secrets: Secrets;

  constructor(http: HttpClient, resolver: Resolver) {
    super(http, resolver, "/api/environments", "environment");
    this.secrets = new Secrets(http, resolver, this.path, this.what);
  }
}

/**
 * The provider accounts this tenant has signed in to, whose credentials
 * Fountain holds (#1178). Connecting one is a browser round trip — send the
 * account owner to `connect_url` from `providers()` — so there is no `create`
 * here. An agent uses one by naming it in `mcp_servers`:
 * `{ gmail: { connection: "<id>" } }`. Only for accounts the egress broker is
 * on for; elsewhere every call is a `NotFoundError` (`connections_not_enabled`).
 */
export class Connections {
  private readonly http: HttpClient;

  constructor(http: HttpClient) {
    this.http = http;
  }

  /** Every connection on the account, active or revoked. Never a token. */
  async list(): Promise<Connection[]> {
    return this.http.list<Connection>("/api/connections");
  }

  /** One connection by id. Unenveloped, as the server answers `show`. */
  async get(id: string): Promise<Connection> {
    return this.http.request<Connection>("GET", `/api/connections/${encodeURIComponent(id)}`);
  }

  /** What this deployment can connect, with the URL that starts each flow. */
  async providers(): Promise<ConnectionProvider[]> {
    return this.http.list<ConnectionProvider>("/api/connections/providers");
  }

  /** Revoke at the provider and delete. Agents that name it get `connection revoked`. */
  async delete(id: string): Promise<void> {
    await this.http.request("DELETE", `/api/connections/${encodeURIComponent(id)}`);
  }
}

/** The vaults on this account, and their secrets. */
export class Vaults extends Collection<Vault, VaultInput, VaultPatch> {
  readonly secrets: Secrets;

  constructor(http: HttpClient, resolver: Resolver) {
    super(http, resolver, "/api/vaults", "vault");
    this.secrets = new Secrets(http, resolver, this.path, this.what);
  }
}

File: sdk/swift/Sources/Fountain/Resources.swift
import Foundation

public final class ResourceCollection: @unchecked Sendable {
  let http: FountainHTTPClient
  let resolver: ResourceResolver
  let path: String
  let what: String

  init(http: FountainHTTPClient, resolver: ResourceResolver, path: String, what: String) {
    self.http = http
    self.resolver = resolver
    self.path = path
    self.what = what
  }

  public func list(search: String? = nil) async throws -> [JSONObject] {
    try await http.list(path, query: ["search": search])
  }

  public func get(_ nameOrID: String) async throws -> JSONObject {
    let id = try await resourceID(nameOrID)
    return try await http.data("GET", "\(path)/\(id)")
  }

  public func create(_ input: JSONObject) async throws -> JSONObject {
    let output = try await http.data("POST", path, body: input)
    await resolver.forget(path)
    return output
  }

  public func update(_ nameOrID: String, patch: JSONObject) async throws -> JSONObject {
    let output = try await http.data(
      "PATCH", "\(path)/\(try await resourceID(nameOrID))", body: patch)
    await resolver.forget(path)
    return output
  }

  public func delete(_ nameOrID: String) async throws {
    _ = try await http.request("DELETE", "\(path)/\(try await resourceID(nameOrID))")
    await resolver.forget(path)
  }

  func resourceID(_ nameOrID: String) async throws -> String {
    guard
      let id = try await resolver.resolve(path: path, what: what, nameOrID: nameOrID)["id"]?
        .stringValue
    else {
      throw FountainError(.resolution, "Resolved \(what) did not include an id")
    }
    return id
  }
}

public final class Secrets: @unchecked Sendable {
  private let collection: ResourceCollection
  init(collection: ResourceCollection) { self.collection = collection }

  public func list(_ parent: String) async throws -> [JSONObject] {
    try await collection.http.list(
      "\(collection.path)/\(try await collection.resourceID(parent))/secrets")
  }

  public func set(_ parent: String, key: String, value: String) async throws -> JSONObject {
    try await collection.http.data(
      "POST", "\(collection.path)/\(try await collection.resourceID(parent))/secrets",
      body: ["key": .string(key), "value": .string(value)]
    )
  }

  public func setAll(_ parent: String, secrets: [String: String]) async throws -> [JSONObject] {
    var output: [JSONObject] = []
    for key in secrets.keys.sorted() {
      output.append(try await set(parent, key: key, value: secrets[key]!))
    }
    return output
  }

  public func delete(_ parent: String, key: String) async throws {
    _ = try await collection.http.request(
      "DELETE",
      "\(collection.path)/\(try await collection.resourceID(parent))/secrets/\(key.pathEncoded)"
    )
  }
}

public final class Environments: @unchecked Sendable {
  private let collection: ResourceCollection
  public let secrets: Secrets
  init(http: FountainHTTPClient, resolver: ResourceResolver) {
    let collection = ResourceCollection(
      http: http, resolver: resolver, path: "/api/environments", what: "environment")
    self.collection = collection
    self.secrets = Secrets(collection: collection)
  }
  public func list(search: String? = nil) async throws -> [JSONObject] {
    try await collection.list(search: search)
  }
  public func get(_ value: String) async throws -> JSONObject { try await collection.get(value) }
  public func create(_ input: JSONObject) async throws -> JSONObject {
    try await collection.create(input)
  }
  public func update(_ value: String, patch: JSONObject) async throws -> JSONObject {
    try await collection.update(value, patch: patch)
  }
  public func delete(_ value: String) async throws { try await collection.delete(value) }
}

public final class Vaults: @unchecked Sendable {
  private let collection: ResourceCollection
  public let secrets: Secrets
  init(http: FountainHTTPClient, resolver: ResourceResolver) {
    let collection = ResourceCollection(
      http: http, resolver: resolver, path: "/api/vaults", what: "vault")
    self.collection = collection
    self.secrets = Secrets(collection: collection)
  }
  public func list(search: String? = nil) async throws -> [JSONObject] {
    try await collection.list(search: search)
  }
  public func get(_ value: String) async throws -> JSONObject { try await collection.get(value) }
  public func create(_ input: JSONObject) async throws -> JSONObject {
    try await collection.create(input)
  }
  public func update(_ value: String, patch: JSONObject) async throws -> JSONObject {
    try await collection.update(value, patch: patch)
  }
  public func delete(_ value: String) async throws { try await collection.delete(value) }
}

public typealias Agents = ResourceCollection

public final class Connections: @unchecked Sendable {
  private let http: FountainHTTPClient
  public let providers: ConnectionProviders
  init(http: FountainHTTPClient) {
    self.http = http
    self.providers = ConnectionProviders(http: http)
  }
  public func list() async throws -> [JSONObject] { try await http.list("/api/connections") }
  public func get(_ id: String) async throws -> JSONObject {
    try await http.request("GET", "/api/connections/\(id.pathEncoded)").objectValue ?? [:]
  }
  public func delete(_ id: String) async throws {
    _ = try await http.request("DELETE", "/api/connections/\(id.pathEncoded)")
  }
}

public final class ConnectionProviders: @unchecked Sendable {
  private let http: FountainHTTPClient
  init(http: FountainHTTPClient) { self.http = http }
  public func list() async throws -> [JSONObject] {
    try await http.list("/api/connection-providers")
  }
  public func get(_ id: String) async throws -> JSONObject { try await raw("GET", id) }
  public func create(_ input: JSONObject) async throws -> JSONObject {
    try await http.request("POST", "/api/connection-providers", body: .object(input)).objectValue
      ?? [:]
  }
  public func update(_ id: String, patch: JSONObject) async throws -> JSONObject {
    try await raw("PATCH", id, body: patch)
  }
  public func delete(_ id: String) async throws {
    _ = try await http.request("DELETE", "/api/connection-providers/\(id.pathEncoded)")
  }
  public func discover(_ id: String) async throws -> JSONObject {
    try await http.request("POST", "/api/connection-providers/\(id.pathEncoded)/discover")
      .objectValue ?? [:]
  }
  private func raw(_ method: String, _ id: String, body: JSONObject? = nil) async throws
    -> JSONObject
  {
    try await http.request(
      method, "/api/connection-providers/\(id.pathEncoded)", body: body.map(JSONValue.object)
    ).objectValue ?? [:]
  }
}
